### PR TITLE
add 2x faster shuffle algorithm; add shuffle bench/plot

### PR DESCRIPTION
**Pull request by knighton**
## Description of changes:
- Improved the speed of shuffling algorithm by 2x
  - **py2s(old approach)**: We generated the sample IDs in the beginning, and did a fixed intra-shard shuffle before partitioning over canonical nodes (ie, splitting some shards) to be "correct". Hence, we were doing shuffling twice. 
  - **py1s(new approach)**: We generate the intra-shard shuffle only once to speed up the computation. In practice, we also found this does matter to convergence, while making us faster.
- Added a shuffling benchmarking script and plot for comparison

Algorithm was tested to convergence.

### CODE REVIEW

**Comment by knighton** [comment]
plot

**Review by dakinggg — COMMENTED** [review]
Can you explain what makes this faster? And the implications on how this changes the "shuffledness" of the shuffle?

**Comment by knighton** [comment]
> Can you explain what makes this faster? And the implications on how this changes the "shuffledness" of the shuffle?

The high bit of the work being done in both algos is shuffling samples within each shard.

Originally, we generated the sample IDs in the beginning, and did a fixed intra-shard shuffle before partitioning over canonical nodes (ie, splitting some shards) to be "correct".

This was found to not really matter. Now we only shuffle intra-shard once, at the end, instead of twice. See plot to verify is 2x.

**Review by dakinggg — COMMENTED** [review]
This looks good to me in terms of correctly implementing what it claims to (with the exception of my one comment). 

1) Is there any way we can test this for correctness? Like maybe at least verify that it doesn't crash or drop or add any ids for a variety of edge casey shard sizes.
2) Is there any way we can verify that this shuffle is "good enough"? Do we need to just try to train some models with it and see what happens? Or is there another way?

File: streaming/base/shuffle.py [hunk 1]
```diff
@@ -138,3 +138,108 @@ def get_shuffle(shard_sizes: NDArray[np.int64], num_canonical_nodes: int, seed:
...
+
+            out_spans.append(span)
+            span_index += 1
+            samples_so_far += samples_this_span
+
+        super_span = begin_part, len(out_spans)
+        super_spans.append(super_span)
+        begin_part = len(out_spans)
+
+    return spans, super_spans
```
[the remark this answers is not in the document]
Comment (knighton) [reply streaming/base/shuffle.py]: fixed

File: scripts/shuffle/plot.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def parse_args() -> Namespace:
+    """Parse command-line arguments.
+
+    Returns:
+        Namespace: Command-line arguments.
+    """
+    args = ArgumentParser()
+    args.add_argument('--in', type=str, required=True)
```
Comment (karan6181) [inline scripts/shuffle/plot.py:36]: Please add a `help` section on what `--in` and `--out` means. Thanks!

File: streaming/base/shuffle/py1s.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Shuffling algorithm that shuffles intra-shard in one place.
+
+This algorithm is slightly more biased and also faster than algorithm ``py2s``.
```
Comment (karan6181) [inline streaming/base/shuffle/py1s.py:6]: Can you please explain what do you mean by `This algorithm is slightly more biased ` ?

**Review by karan6181 — COMMENTED** [review]
Overall LGTM. Few minor comments

- Can you please add more details in the description on why new shuffling algorithm is faster ?
- Can you please also list down the end-to-end integration test that you did?

File: scripts/epoch/bench.py [hunk 4]
```diff
@@ -21,6 +21,7 @@ def parse_args() -> Namespace:
     """
     args = ArgumentParser()
     args.add_argument('--index', type=str, required=True)
+    args.add_argument('--shuffle_algo', type=str, default='py2s')
```
[the remark this answers is not in the document]
Comment (knighton) [reply scripts/epoch/bench.py:24]: done

File: scripts/shuffle/bench.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        dataset_size (int): Total samples in the dataset.
+        samples_per_shard (float): Average shard size in samples.
+
+    Returns:
+        NDArray[np.int64]: The sample size of each shard.
+    """
+    num_shards = math.ceil(dataset_size / samples_per_shard)
+    shard_sizes = num_shards * [samples_per_shard]
+    shard_sizes[-1] -= num_shards * samples_per_shard - dataset_size
+    assert sum(shard_sizes) == dataset_size
```
[the remark this answers is not in the document]
Comment (knighton) [reply scripts/shuffle/bench.py:87]: removing this line, as the math is correct

File: scripts/shuffle/bench.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        return Caller(func, args.num_canonical_nodes, args.seed, args.epoch, args.timeout)
+
+    callers = list(map(wrap, get_shuffles))
+
+    text = ' '.join(map(lambda s: s.rjust(10), names))
+
+    print(f'{"power".rjust(5)} {"samples".rjust(14)} ' + text)
+    for mul_power in range(args.min_power * args.power_interval,
+                           args.max_power * args.power_interval + 1):
+        power = mul_power / args.power_interval
```
[the remark this answers is not in the document]
Comment (knighton) [reply scripts/shuffle/bench.py:110]: powers here are intentionally fractional, done in this weird way to avoid accumulating floating point error

File: scripts/shuffle/plot.py (hunk 2 above)
Comment (knighton) [reply scripts/shuffle/plot.py:36]: done

File: streaming/base/shuffle/py1s.py (hunk 3 above)
Comment (knighton) [reply streaming/base/shuffle/py1s.py:6]: added explanations

**Review by karan6181 — APPROVED** [review]

**Comment by knighton** [comment]
Screenshot 2023-02-23 at 3 05 12 PM

Convergence tests are looking good.

File: streaming/base/shuffle/py1s.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Shuffling algorithm that shuffles intra-shard in one place.
+
+This algorithm is roughly twice as fast as algorithm ``py2s``, and ever so slightly biased.
+
+Bias in this case merely refers to how we assign samples when we split shards at canonical node
+boundaries, which is non-random in this algorithm. In practice, we found this does matter to
```
Comment (dakinggg) [inline streaming/base/shuffle/py1s.py:9]: is this supposed to say "does not matter"

Comment (knighton) [reply streaming/base/shuffle/py1s.py:9]: https://github.com/mosaicml/streaming/pull/173
